### PR TITLE
Add query_table_version to the rest client

### LINKED ISSUE #82: Expose table version in metadata to make polling for [data] updates faster & easier
Looking for table version # (or timestamp) so that we can build an airflow sensor to easily detect if a new version of the data is now available.

image

### PR DESCRIPTION
**Pull request by nkvuong**
This is proposed in #82 

Also remove the restriction for Python 3.10 #110 

Closes #82 and #110

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -106,3 +106,12 @@ spark-warehouse/
...
 *.egg-info
+
+# For VSCode
+*.vscode
+
+# For Metals
+*.metals
+
+# For venv
+*.venv
```
Comment (zhuansunxt) [inline .gitignore:117]: pls have a new line at EOF

File: python/delta_sharing/rest_client.py [hunk 2]
```diff
@@ -253,6 +266,15 @@ def _get_internal(self, target: str, data: Optional[Dict[str, Any]] = None):
     def _post_internal(self, target: str, data: Optional[Dict[str, Any]] = None):
         return self._request_internal(request=self._session.post, target=target, json=data)
 
+    def _head_internal(self, target: str):
+        assert target.startswith("/"), "Targets should start with '/'"
```
Comment (zhuansunxt) [inline python/delta_sharing/rest_client.py:270]: Can we do what other two methods are doing, e.g `return self._request_internal(request=self._session.head, target=target)`

File: python/delta_sharing/tests/test_rest_client.py [hunk 3]
```diff
@@ -205,6 +205,14 @@ def test_query_table_metadata_partitioned_different_schemas(rest_client: DataSha
     )
 
 
+@pytest.mark.skipif(not ENABLE_INTEGRATION, reason=SKIP_MESSAGE)
+def test_query_table_version(rest_client: DataSharingRestClient):
+    response = rest_client.query_table_version(
+        Table(name="table1", share="share1", schema="default")
+    )
+    assert response.delta_table_version == 28
```
Comment (zhuansunxt) [inline python/delta_sharing/tests/test_rest_client.py:213]: This hardcoded value is decided by the current version of the table and if that table is updated this test will break, so maybe just assert some integer value is defined.

File: python/delta_sharing/rest_client.py (hunk 2 above)
Comment (nkvuong) [reply python/delta_sharing/rest_client.py:271]: `_request_internal` returns only the request content, but we only need the headers from the request, hence a new function

File: python/delta_sharing/tests/test_rest_client.py (hunk 3 above)
Comment (nkvuong) [reply python/delta_sharing/tests/test_rest_client.py]: added int assertion and that table version is positive

File: python/setup.py [hunk 4]
```diff
@@ -40,7 +40,7 @@
     packages=[
         'delta_sharing',
     ],
-    python_requires='>=3.6,<3.10',
+    python_requires='>=3.6,<=3.10',
```
Comment (zsxwing) [inline python/setup.py:43]: Could you remove `,<=3.10` to make us support future Python versions?

File: python/delta_sharing/rest_client.py [hunk 5]
```diff
@@ -219,6 +224,14 @@ def query_table_metadata(self, table: Table) -> QueryTableMetadataResponse:
                 metadata=Metadata.from_json(metadata_json["metaData"]),
             )
 
+    @retry_with_exponential_backoff
+    def query_table_version(self, table: Table) -> QueryTableVersionResponse:
+        headers = self._head_internal(
+            f"/shares/{table.share}/schemas/{table.schema}/tables/{table.name}"
+        )
+        table_version = int(headers.get("delta-table-version", None) or -1)
```
Comment (zsxwing) [inline python/delta_sharing/rest_client.py:232]: Could you raise an error when the header is missing? This is a required header defined in the protocol. If a server doesn't return it, it should be a bug in the server and it's better to raise an error.

Comment (nkvuong) [reply python/delta_sharing/rest_client.py:232]: if the table does not exist on the server, should the server still return a delta-table-version in the header? Happy to raise a LookupError when the header is missing

File: .gitignore (hunk 1 above)
Comment (nkvuong) [reply .gitignore]: done

File: python/delta_sharing/rest_client.py (hunk 5 above)
Comment (zsxwing) [reply python/delta_sharing/rest_client.py:232]: Good point. This reminds me that we are missing `response.raise_for_status()` in `_head_internal`. We should fail if the server turns non 200 response. Could you also add a test for this?

**Review by zsxwing — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**